### PR TITLE
Add configuration file reference page

### PR DESCRIPTION
**Pull request by jbogarthyde**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
Add reference documentation for angular.json configuration options.

### CODE REVIEW

**Review by alan-agius4 — CHANGES_REQUESTED** [review]
Couple of overall comments.

1) Options should always be camelCase
2) Most of the builders have missing options. Pretty much, all options in an `ng` command excluding `--help` should be included together with any addition options. These can be found here https://github.com/angular/angular-cli/tree/master/packages/angular_devkit/build_angular/src
3) There are some missing builders, such as `App-Shell`, `Universal`, `E2E`, etc...

**Review by hansl — APPROVED** [review]
LGTM after @alan-agius4 's comments have been fixed (doesn't look like it).

File: aio/content/guide/config-ref.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+{@a build-props}
+
+### Additional build and test options
+
+The configurable options for a default or targeted build correspond to the options available for the [`ng build`](https://angular.io/cli/build) and [`ng test`](https://angular.io/cli/test) commands, with the following additions.
+
+| OPTIONS PROPERTIES | DESCRIPTION |
+| :------------------------- | :---------------------------- |
+| `fileReplacements`         | An array of files and their compile-time replacements. |
+| `stylePreprocessorOptions` | An array of option-value pairs to pass to style preprocessors. |
```
Comment (alan-agius4) [inline aio/content/guide/config-ref.md]: stylePreopcessorOptions is not array, it’s child in an array. Thus I suggest to change `An Array` to `An Object`

Same for the same property in other builders

File: aio/content/guide/config-ref.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  app_name-e2e
+    ...
+</code-example>
+
+Each additional app that you create with `ng generate application` has a corresponding end-to-end test project, with its own configuration section.
+When you create a library project with `ng generate library`, the library project is also added to the `projects` section. 
+
+<div class="alert is-helpful">
+  Note that the `projects` section of the configuration file does not correspond exactly to the workspace file structure. 
+  * The initial app created by `ng new` is at the top level of the workspace file structure, along with its e2e app.
```
Comment (alan-agius4) [inline aio/content/guide/config-ref.md]: Can you confirm if the `*` looks good in the helpful alerts?

File: aio/content/guide/config-ref.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    }
+</code-example>
+
+| PROPERTY | DESCRIPTION |
+| :-------------- | :---------------------------- |
+| `root`          | The root folder for this project's files, relative to the workspace folder. Empty for the initial app, which resides at the top level of the workspace. |
+| `sourceRoot`    | The root folder for this project's source files. |
+| `projectType`   | One of "application" or "library". An application can run independently in a browser, while a library cannot. Both an app and its e2e test app are of type "application".|
+| `prefix`        | A string that Angular prepends to generated selectors. Can be customized to identify an app or feature area. |
+| `schematics`    | An array of schematics that customize CLI commands for this project. |
```
Comment (alan-agius4) [inline aio/content/guide/config-ref.md]: Schematics and architect are not an array technically it's an object.

Wdyt @hansl ?

**Review by alan-agius4 — COMMENTED** [review]
Couple of small NITs

File: aio/content/guide/config-ref.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+* The `architect/build` section configures defaults for options of the `ng build` command.
+
+* The `architect/serve` section overrides build defaults and supplies additional serve defaults for the `ng serve` command.  In addition to the options available for the `ng build` command, it adds options related to serving the app.
+
+* The `architect/test` section overrides build-option defaults for test builds and supplies additional test-running defaults for the `ng test` command.
+
+* The `architect/lint` section configures defaults for options of the `ng lint` command, which performs code analysis on project source files.  The default linting tool for Angular is [TSLint](https://palantir.github.io/tslint/).
+
+* The `architect/extract-i18n` section configures defaults for options of the `ng-xi18n` tool used by the `ng xi18n` command, which extracts marked message strings from source code and outputs translation files.
```
Comment (alan-agius4) [inline aio/content/guide/config-ref.md]: Is it on purpose that E2E and universal builders are not listed here.

**Review by alan-agius4 — APPROVED** [review]
Thanks! LGTM

File: aio/content/guide/config-ref.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+* The `architect/e2e` section overrides build-option defaults for building end-to-end testing apps with the `ng e2e` command.
+
+* The `architect/test` section overrides build-option defaults for test builds and supplies additional test-running defaults for the `ng test` command.
+
+* The `architect/lint` section configures defaults for options of the `ng lint` command, which performs code analysis on project source files.  The default linting tool for Angular is [TSLint](https://palantir.github.io/tslint/).
+
+* The `architect/extract-i18n` section configures defaults for options of the `ng-xi18n` tool used by the `ng xi18n` command, which extracts marked message strings from source code and outputs translation files.
+
+* The `architect/universal` section configures defaults for building an app with server-side rendering with the `ng universal` command.
```
Comment (alan-agius4) [inline aio/content/guide/config-ref.md]: At this time there is no `ng universal` at the moment users need to run `ng run <project>:universal:<configuration>`

**Review by alan-agius4 — COMMENTED** [review]
Two small comments

File: aio/content/navigation.json [hunk 6]
```diff
@@ -515,6 +515,12 @@
           "title": "Project File Structure",
           "tooltip": "How your Angular workspace looks in your filesystem."
         },
+        {
+          "url": "guide/config-ref",
```
Comment (IgorMinar) [inline aio/content/navigation.json]: could we change the url to be something more memorable as "cli-config" please? (this will also require file rename of the md file.

Comment (IgorMinar) [reply aio/content/navigation.json]: or since the title of the doc is "Workspace configuration" then maybe "guide/workspace-config" would be better?

File: aio/content/navigation.json [hunk 7]
```diff
@@ -515,6 +515,12 @@
           "title": "Project File Structure",
           "tooltip": "How your Angular workspace looks in your filesystem."
```
Comment (IgorMinar) [inline aio/content/navigation.json]: a drive-by correction: oops.. there is a typo here. it should be "on your filesystem"

File: aio/content/guide/config-ref.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
+# Angular Workspace Configuration
+
+A file named `angular.json` at the root level of an Angular [workspace](guide/glossary#workspace) provides workspace-wide and project-specific configuration defaults.
```
Comment (IgorMinar) [inline aio/content/guide/config-ref.md]: we didn't mention what is this configuration for... maybe add "configuration defaults for build and development tools provided by the Angular CLI"?

File: aio/content/guide/config-ref.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+Each additional app that you create with `ng generate application` has a corresponding end-to-end test project, with its own configuration section.
+When you create a library project with `ng generate library`, the library project is also added to the `projects` section. 
+
+<div class="alert is-helpful">
+
+  Note that the `projects` section of the configuration file does not correspond exactly to the workspace file structure. 
+  * The initial app created by `ng new` is at the top level of the workspace file structure, along with its e2e app.
+  * Additional apps, e2e apps, and libraries go into a `projects` folder in the workspace.
+
```
Comment (IgorMinar) [inline aio/content/guide/config-ref.md]: you could also link to the workspace file structure layout doc for more info...

File: aio/content/guide/config-ref.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+You can define and name additional alternate configurations (such as `stage`, for instance) appropriate to your development process. Some examples of different build configurations are `stable`, `archive` and `next` used by AIO itself, and the individual locale-specific configurations required for building localized versions of an app. For details, see [Internationalization (i18n)](guide/i18n#merge-aot). 
+
+{@a build-props}
+
+### Additional build and test options
+
+The configurable options for a default or targeted build correspond to the options available for the [`ng build`](https://angular.io/cli/build) and [`ng test`](https://angular.io/cli/test) commands, with the following additions.
+
+| OPTIONS PROPERTIES | DESCRIPTION |
```
Comment (IgorMinar) [inline aio/content/guide/config-ref.md]: for all of these properties we need to provide the possible values, or examples of values, otherwise I just know that these properties exist, but I don't know what I can set the values to.

File: aio/content/guide/config-ref.md [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        "e2e" : { },
+        "test": { },
+        "lint": { },
+        "extract-i18n": { },
+        "server": { },
+        "app-shell": { }
+      }
+</code-example>
+
+* The `architect/build` section configures defaults for options of the `ng build` command.
```
Comment (IgorMinar) [inline aio/content/guide/config-ref.md]: + "see [Build target] section for more info"

File: aio/content/guide/config-ref.md [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+| :------------------------- | :---------------------------- |
+| `fileReplacements`         | An object containing files and their compile-time replacements. |
+| `stylePreprocessorOptions` | An object containing option-value pairs to pass to style preprocessors. |
+| `assets`                   | An object containing paths to static assets to add to the global context of the project. The default paths point to the project's icon file and its `assets` folder. |
+| `styles`                   | An object containing style files to add to the global context of the project. Angular CLI supports CSS imports and all major CSS preprocessors: [sass/scss](http://sass-lang.com/), [less](http://lesscss.org/), and [stylus](http://stylus-lang.com/). |
+| `scripts`                  | An object containing JavaScript script files to add to the global context of the project. The scripts are loaded exactly as if you had added them in a `<script>` tag inside `index.html`. |
+| `budgets`                  | Default size-budget threshholds for all or parts of your app. See [Configure size budgets](guide/build#configure-size-budgets). (Not available in `test` section.)|
```
Comment (IgorMinar) [inline aio/content/guide/config-ref.md]: There are many other properties that are not listed in this table. e.g. a new app usually has this prod configuration:

```
"configurations": {
            "production": {
              "fileReplacements": [
                {
                  "replace": "src/environments/environment.ts",
                  "with": "src/environments/environment.prod.ts"
                }
              ],
              "optimization": true,
              "outputHashing": "all",
              "sourceMap": false,
              "extractCss": true,
              "namedChunks": false,
              "aot": true,
              "extractLicenses": true,
              "vendorChunk": false,
              "buildOptimizer": true,
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "2mb",
                  "maximumError": "5mb"
                }
            }
```

and most of the properties in this config are missing in the table above. can we add them please?

**Review by IgorMinar — CHANGES_REQUESTED** [review]
This doc is on the right path, but needs a bit more work.. thanks!

File: aio/content/guide/config-ref.md (hunk 12 above)
Comment (jbogarthyde) [reply aio/content/guide/config-ref.md]: Alan and I talked about this. The config options that correspond to command options are already fully documented in the CLI command reference (or soon will be). If we try to repeat that doc here, we have to be sure it agrees, and we run into a maintenance problem.

Alan and I thought it would be better to reference the CLI command reference where possible, and only provide additional doc here for options that can ONLY be set through the config file. That's what I've tried to do. I will try to make it more explicit.

File: aio/content/guide/workspace-config.md [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+| :-------------- | :---------------------------- |
+| `builder`       | The npm package for the build tool used to create this target. The default is `@angular-devkit/build-angular:browser`, which uses the [webpack](https://webpack.js.org/) package bundler. |
+| `options`       | This section contains defaults for build options, used when no named alternative configuration is specified. See [Default build options](#build-props) below. |
+| `configurations`| This section defines and names alternative configurations for different intended destinations. It contains a section for each named configuration, which sets the default options for that intended environment. See [Alternate build configurations](#build-configs) below. |
+
+{@a build-configs}
+
+### Alternate build configurations
+
+By default, a `production` configuration is defined, and the `ng build` command has `--prod` option that builds using this configuration. The `production` configuration sets defaults that optimize the app in a number of ways, such bundling files, minimizing excess whitespace, removing comments and dead code, and rewriting code to use short, cryptic names ("uglification").
```
Comment (IgorMinar) [inline aio/content/guide/workspace-config.md]: "uglification" refers to "Uglify" which is a product name. "minification" is a general term that would be more suitable because we actually no longer use "Uglify" in our toolchain to do minification.

**Review by IgorMinar — APPROVED** [review]
one small correction. but otherwise lgtm. thanks
